### PR TITLE
8342086: FileInputStream.available() fails with "Incorrect function" for "nul" path (win)

### PR DESCRIPTION
**Pull request by bplb**
Modify the `handleAvailable` native code path to ignore an `ERROR_INVALID_FUNCTION` encountered by `PeekNamedPipe` and update an existing test to verify this change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342086](https://bugs.openjdk.org/browse/JDK-8342086): FileInputStream.available() fails with "Incorrect function" for "nul" path (win) (**Bug** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21543/head:pull/21543` \
`$ git checkout pull/21543`

Update a local copy of the PR: \
`$ git checkout pull/21543` \
`$ git pull https://git.openjdk.org/jdk.git pull/21543/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21543`

View PR using the GUI difftool: \
`$ git pr show -t 21543`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21543.diff">https://git.openjdk.org/jdk/pull/21543.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21543#issuecomment-2417412000)
</details>

### CODE REVIEW

**Comment by bplb** [comment]
If `PeekNamedPipe` fails with `ERROR_INVALID_FUNCTION`, then the number of bytes available is set to zero, the default. The test is updated to JUnit 5 and a sub-test is added for this case of reading form the `NUL` device.

**Comment by bplb** [comment]
I verified that `PeekNamedPipe` also fails with `ERROR_INVALID_FUNCTION` when the supplied handle is `STD_INPUT_HANDLE`.

**Comment by bplb** [comment]
continue;

**Comment by bplb** [comment]
Ping ...

File: src/java.base/windows/native/libjava/io_util_md.c [hunk 1]
```diff
@@ -346,10 +346,14 @@ handleNonSeekAvailable(FD fd, long *pbytes) {
...
+         * The only alternative is to raise an Exception,
+         * which isn't really warranted. It also fails if the
+         * handle is to the NUL device (ERROR_INVALID_FUNCTION),
+         * so *pbytes = 0 is also set in that case.
          */
-        if (GetLastError() != ERROR_BROKEN_PIPE) {
+        DWORD lastError = GetLastError();
+        if (lastError != ERROR_BROKEN_PIPE &&
+            lastError != ERROR_INVALID_FUNCTION) {
             return FALSE;
```
[the remark this answers is not in the document]
Comment (bplb) [reply src/java.base/windows/native/libjava/io_util_md.c:357]: > [...] not special case these errors and instead pbytes=0 and TRUE if PeekNamedPipe fails.

Sounds reasonable.

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**